README
                               Text::BibTeX
                               version 0.34
                       Greg Ward (gward@python.net)
                             22 December, 2000

INTRODUCTION
------------

Text::BibTeX is a Perl library for reading, parsing, and processing
BibTeX files.  It is the Perl half of btOOL, a pair of libraries for
dealing with BibTeX data; the C half, btparse, is needed to compile
Text::BibTeX (see below).

Text::BibTeX gives you access to the data at many different levels: you
may work with BibTeX entries as simple "field -> string" mappings, or
get at the original form of the data as a list of simple values
(strings, macros, or numbers) pasted together.  You can choose not to
impose nany restrictions on the allowed/expected entry types or fields,
or you can use the structure defined by BibTeX 0.99's standard style
files, or you can invent your own.

The library is copiously documented.  After installing the module, see
the Text::BibTeX man page for a comprehensive introduction to the
system.  If you wish to dive straight in without regards for niceties
such as splitting and formatting names, defining or imposing database
structures, getting access to unprocessed field values, and other nifty
features, then you can probably get away with just reading the
Text::BibTeX::Entry man page.  Comments/criticism of the documentation
are welcome.

In addition to the man pages embedded in each module (and available
after installation), I have written a technical report describing btOOL,
with the btparse and Text::BibTeX documentation included as appendices.
The whole report is just over 100 pages, around 45 of which make up the
Text::BibTeX documentation (the btparse documentation is a further 30
pages).  You can find it at the btOOL home page:

    http://starship.python.net/~gward/btOOL/

I may also make the btparse and Text::BibTeX manuals available as
separate PostScript files, but they aren't there as of this writing.


DEPENDENCIES
------------

Text::BibTeX requires Perl 5.004 or later and an ANSI-compliant C
compiler.

You must also have the btparse distribution, the C library underlying
Text::BibTeX.  If you didn't download btparse with Text::BibTeX, don't
sweat it; the Makefile.PL is smart enough to go out and find it for you
(using your preferred CPAN mirror if you have set one up,
http://www.cpan.org/ otherwise).  If you did download btparse, just put
it in the same directory as the Text::BibTeX tarball.  Text::BibTeX's
Makefile.PL will unpack and configure it for you.


BUILDING
--------

Start by generating the Makefile and other preparatory steps (most notably,
preparing the btparse package for building).

   perl Makefile.PL

This goes to some lengths to ensure that btparse, the companion C library,
is found and available.  It will configure btparse for you, so you
shouldn't have to do anything in the 'btparse' directory unless things go
wrong there.

Note that if Perl was built with a non-ANSI C compiler, you will need to
specify an ANSI compiler when running Makefile.PL.  For instance, if you're
running SunOS and Perl was built with 'cc' (non-ANSI), you will need to
build Text::BibTeX (and btparse as well, for that matter) with an ANSI
compiler such as 'gcc' or Sun's 'acc'.  This can be done for Text::BibTeX
as follows:

   perl Makefile.PL CC=gcc      # or acc, or whatever works on your system

However, this is fraught with danger and not recommended.  The only safe
thing to do is to build Perl with an ANSI-compliant compiler.  This will
probably save you trouble with other extensions that require an ANSI C
compiler.

Now, build everything:

   make

This takes care of building btparse as well as the Text::BibTeX modules, so
again you don't have to go grubbing around in the btparse directory unless
things go wrong.

Optionally, you can test and install btparse:

   cd btparse
   make test
   make install
   cd ..

(You do not have to install btparse to use Text::BibTeX; that's only needed
if you plan to do C programming using btparse.  See btparse/README for
details.)

Finally, test and install the Text::BibTeX Perl modules:

   make test
   make install

Note that you will need to be root on most systems in order to install
either btparse or Text::BibTeX.  Consult the 'perlmodinstall' man page (as
of Perl 5.005 or later) for more information on building and installing
Perl module distributions.

Please let me know if anything goes wrong with building either btparse
or Text::BibTeX.


BUGS AND LIMITIATIONS
---------------------

See the btparse(3) and Text::BibTeX(3) man pages for details.  Executive
summary:

  * Text::BibTeX is not thread-safe, due to a heavy dependence on
    global variables in the parser and lexical scanner components
    of btparse

  * for the same reason, you can't have multiple files open and
    being parsed at the same time; attempting to do so is an
    invalid use of btparse

  * Text::BibTeX cannot be used with an sfio-based Perl, because
    the Perl I/O API is apparently not yet up to the task of replacing
    stdio in third-party C libraries


AVAILABILITY
------------

You can find the latest version of Text::BibTeX, as well as
documentation, information on known bugs, etc. at the btOOL web site:

   http://starship.python.net/~gward/btOOL/

Here you will find HTML versions of the documentation and technical
report describing the project, links to download the code, and whatever
other goodies I can come up with over time.


CREDITS
-------

For spotting bugs and sometimes even providing patches:

    Dirk Vleugels <vleugels@do.isst.fhg.de>
    Kjetil Kjernsmo <kjetil.kjernsmo@astro.uio.no>
    Andrew Cassin <acassin@cs.mu.oz.au>
    Thomas Kamphusmann <thomas@delphi.ping.de>
    Horst Tellioglu <telliogl@h2hobel.phl.univie.ac.at>
    Nikolay Pelov <Nikolay.Pelov@cs.kuleuven.ac.be>

...and probably a few others.  Please send me email if you feel you
belong on this list and I forgot to include you.

Also, see the btparse README file for contributors to the C library.

$Id: README 694 2000-12-23 03:55:10Z greg $


CHANGES

Release 0.63 - 12 May, 2012
---------------------------
* Patch on what to consider whitespaces (thanks to Philip Kime)

Release 0.62 - 11 January, 2012
-------------------------------
* Releasing stable

Release 0.62_01 - 7 January, 2012
---------------------------------
* Better installation under Windows, and specifically, Win 7.

Release 0.61 - 20 October, 2011
-------------------------------
* Thanks to Philip Kime (again), we have STDERR working on Windows
  (hopefully)
* Make Windows tests no longer ignored.

Release 0.60 - 31 July, 2011
----------------------------
* Fix test in t/output.t

Release 0.59 - 28 July, 2011
----------------------------
* Add comma at the end of each and any line in BibTeX record
  as all modern parsers support them.
* Rewrote some tests in t/output.t

Release 0.58 - 21 June, 2011
----------------------------
* Remove some old documentation about btparse.
* Warn users when installing on non standard library path.

Release 0.57 - 17 June, 2011
----------------------------
* Mark two tests that fail from time to time as TODO for now.

Release 0.56 - 8 June, 2011
---------------------------
* Use File::Temp instead of POSIX. Fixes some Win32 builds.
  Thanks to Alex Gough for reporting this build problem.

Release 0.55 - 25 April, 2011
-----------------------------
* Added some extra checks for lib64 dir.
* Rewrote some code on MyBuilder.pm
* Add soname to library build.

Release 0.54 - 17 April, 2011
-----------------------------
* Applied patch to work under lib64 archs (hopefuly)
 (Guillaume Rousse)

Release 0.53 - 10 March, 2011
-----------------------------
* Working under Cygwin (Philip Kime)
* Changed Windows testing mode (based on Cygwin approach)
  - getting rid of 00_ and zz_ test files. Great!

Release 0.52 - 15 February, 2011
--------------------------------
* More name abbreviation tweeks (Philip Kime)

Release 0.51 - 1 February, 2011
-------------------------------
* Fixed environment during Linux/Unix testing

Release 0.50 - 1 February, 2011
-------------------------------
* Fixed name abbreviation when name has hiffens (Philip Kime)

Release 0.49 - 27 January, 2011
-------------------------------
* Small patch by Philip Kime to enlarge macros buffer size.
* Fixed Build.PL to clean some object files.

Release 0.48 - 23 September, 2010
---------------------------------
* depend on ExtUtils::LibBuilder for library compilation.

Release 0.47 - 18 September, 2010
---------------------------------
* fixed a problem with 160 char being considered whitespace on many unixes.
  Thanks, again, to Philip Kime :)

Release 0.46 - 24 August, 2010
------------------------------
* support entry keys in unicode. Thanks to Philip Kime.

Release 0.45 - 8 June, 2010
---------------------------
* fixed ldconfig call under linux, now only if the user is root.
* fixed library installation path when user specifies different
  installbase during install action.
  (thanks to François for both fixes)

Release 0.44 - 9 May, 2010
--------------------------
* RPath information on link - Thanks to Jens Rehsack
* removed dependency on 'UNIVERSAL' as it is now built-in
  (change for perl 5.12 deprecation)

Release 0.43 - 18 March, 2010
-----------------------------
* small fix on warnings issue
* run ldconfig after installing the .so file under linux

Release 0.42 - 16 March, 2010
-----------------------------
* names with more than two commas are left untouched if protected by
  braces.
* can now deal with names with hyphens and generate the correct BibTeX
  compatible abbreviations.
* can generate initials for protected name parts, like BibTeX can.
* can deal with generating initials for names which contains Unicode
  characters.
* fixed compilation on machines needing alloca.h
* added new README file with more up-to-date information.

Release 0.41 - 14 March, 2010
-----------------------------
* tested under Solaris and FreeBSD.
* fixed compilation/installation under Mac OS X.
* fixed some C code on format-security (thanks to Jerome Quelin).
* fixed build system to check for dependencies.
* added patch from Philip Kime to fix names with commas handling.

Release 0.40 - 7 March, 2010
----------------------------
* this release is very similar to 0.40_3, but now indexable by cpan.

Release 0.40_3 - 18 February, 2010
----------------------------------
* fixed Build.PL with correct build_requirements;
* fixed test to use still not installed library;

Release 0.40_2 - 15 February, 2010
----------------------------------
* merged btparse library into Text::BibTeX code. I am sorry for all
  other languages that might be using this library. For them my
  suggestion is that they make Text::BibTeX as a dependency.
* ported compilation and configuration tools from ExtUtils::MakeMaker
  to Module::Build.
* back on track, trying to make Text::BibTeX work and compile easily
  on main platforms (sorry, for main platforms I assume Linux, Mac OS
  X and Windows running Strawberry Perl).
* Thanks to Philip Kime for continuous poke so I work on this!!
  Created a THANKS file for this purpose.

Release 0.38 - 8 March, 2008
----------------------------
* fixed problem when creating an empty Entry.
  Thanks to Frank Wiegand.

Release 0.37 - 21 Setember, 2006
--------------------------------
* added support for 'conference' key (alias for inproceedings)

Release 0.36 - 24 July, 2005
----------------------------
* fixed url where btparse should be downloaded from. (stupid bug)

Release 0.35 - 23 July, 2005
----------------------------
* changed from 0.33 to 0.35 to maintain relation with btparse.
* some tests now use Test::More (more tests will change soon)
* the way the module initializes changed. Now the module can be
  included to check its version.

Release 0.33 - 22 March, 2000
-----------------------------

BibTeX/Bib.pm
# ----------------------------------------------------------------------
# NAME       : BibTeX/Bib.pm
# CLASSES    : Text::BibTeX::BibStructure, Text::BibTeX::BibEntry;
#              loads Text::BibTeX::BibSort and Text::BibTeX::BibFormat
#              for use by BibEntry
# RELATIONS  : BibStructure inherits from Structure
#              BibEntry inherits from BibSort and BibFormat, which
#                both inherit from StructuredEntry
# DESCRIPTION: Implements the "Bib" structure, which provides the
#              same functionality -- though in a completely different
#              context, and much more customizably -- as the standard
#              style files of BibTeX 0.99.
# CREATED    : 1997/09/21, Greg Ward
# MODIFIED   : 
# VERSION    : $Id: Bib.pm 670 2000-03-23 02:08:40Z greg $
# COPYRIGHT  : Copyright (c) 1997-2000 by Gregory P. Ward.  All rights
#              reserved.
# 
#              This file is part of the Text::BibTeX library.  This
#              library is free software; you may redistribute it and/or
#              modify it under the same terms as Perl itself.
# ----------------------------------------------------------------------

=head1 NAME

Text::BibTeX::Bib - defines the "Bib" database structure

=head1 SYNOPSIS

   $bibfile = new Text::BibTeX::File $filename;
   $bibfile->set_structure ('Bib',
                            # Default option values:
                            sortby => 'name',
                            namestyle => 'full'
                            nameorder => 'first',
                            atitle => 1,
                            labels => 'numeric');

   # Alternate option values:
   $bibfile->set_option (sortby => 'year');
   $bibfile->set_option (namestyle => 'nopunct');
   $bibfile->set_option (namestyle => 'nospace');
   $bibfile->set_option (nameorder => 'last');
   $bibfile->set_option (atitle => 0);   
   $bibfile->set_option (labels => 'alpha');   # not implemented yet!

   # parse entry from $bibfile and automatically make it a BibEntry
   $entry = new Text::BibTeX::Entry $bibfile;

   # or get an entry from somewhere else which is hard-coded to be
   # a BibEntry
   $entry = new Text::BibTeX::BibEntry ...;

   $sortkey = $entry->sort_key;
   @blocks = $entry->format;

=head1 DESCRIPTION

(B<NOTE!> Do not believe everything you read in this document.  The
classes described here are unfinished and only lightly tested.  The
current implementation is a proof-of-principle, to convince myself (and
anyone who might be interested) that it really is possible to
reimplement BibTeX 0.99 in Perl using the core C<Text::BibTeX> classes;
this principle is vaguely demonstrated by the current C<Bib*> modules,
but not really proved.  Many important features needed to reimplement
the standard styles of BibTeX 0.99 are missing, even though this
document may brashly assert otherwise.  If you are interested in using
these classes, you should start by reading and grokking the code, and
contributing the missing bits and pieces that you need.)

C<Text::BibTeX::Bib> implements the database structure for
bibliographies as defined by the standard styles of BibTeX 0.99.  It
does this by providing two classes, C<BibStructure> and C<BibEntry> (the
leading C<Text::BibTeX> is implied, and will be omitted for the rest of
this document).  These two classes, being specific to bibliographic
data, are outside of the core C<Text::BibTeX> class hierarchy, but are
distributed along with it as they provide a canonical example of a
specific database structure using classes derived from the core
hierarchy.

C<BibStructure>, which derives from the C<Structure> class, deals with
the structure as a whole: it handles structure options and describes all
the types and fields that make up the database structure.  If you're
interested in writing your own database structure modules, the standard
interface for both of these is described in L<Text::BibTeX::Structure>;
if you're just interested in finding out the exact database structure or
the options supported by the C<Bib> structure, you've come to the right
place.  (However, you may have to wade through a bit of excess verbiage
due to this module's dual purpose: first, to reimplement the standard
styles of BibTeX 0.99, and second, to provide an example for other
programmers wishing to implement new or derived database structure
modules.)

C<BibEntry> derives from the C<StructuredEntry> class and provides
methods that operate on individual entries presumed to come from a
database conforming to the structure defined by the C<BibStructure>
class.  (Actually, to be completely accurate, C<BibEntry> inherits from
two intermediate classes, C<BibSort> and C<BibFormat>.  These two
classes just exist to reduce the amount of code in the C<Bib> module,
and thanks to the magic of inheritance, their existence is usually
irrelevant.  But you might want to consult those two classes if you're
interested in the gory details of sorting and formatting entries from
BibTeX 0.99-style bibliography databases.)

=cut


# first, the "structure class" (inherits from Text::BibTeX::Structure)

package Text::BibTeX::BibStructure;
use strict;
use vars qw(@ISA);
@ISA = qw(Text::BibTeX::Structure);

=head1 STRUCTURE OPTIONS

C<BibStructure> handles several user-supplied "structure options" and
methods for dealing with them.  The options currently supported by the
C<Bib> database structure, and the values allowed for them, are:

=over 4

=item C<sortby>

How to sort entries.  Valid values: C<name> (sort on author names, year,
and title), C<year> (sort on year, author names, and title).  Sorting on
"author names" is a bit more complicated than just using the C<author>
field; see L<Text::BibTeX::BibSort> for details.  Default value: C<name>.

=item C<namestyle>

How to print author (and editor) names: C<full> for unabbreviated first
names, C<abbrev> for first names abbreviated with periods, C<nopunct>
for first names abbreviated with space but no periods, and C<nospace> to
abbreviate without space or periods.  Default value: C<full>.

=item C<nameorder>

The order in which to print names: C<first> for "first von last jr"
order, and C<last> for "von last jr first" order.  Default value:
C<first>.

=item C<atitle_lower>

A boolean option: if true, non-book titles will be changed to "sentence
capitalization:" words following colons and sentence-ending punctuation
will be capitalized, and everything else at brace-depth zero will be
changed to lowercase.  Default value: true.

=item C<labels>

The type of bibliographic labels to generate: C<numeric> or C<alpha>.
(Alphabetic labels are not yet implemented, so this option is currently
ignored.)  Default value: C<numeric>.

=back

Also, several "markup options" are supported.  Markup options are
distinct because they don't change how text is extracted from the
database entries and subsequently mangled; rather, they supply bits of
markup that go around the database-derived text.  Markup options are
always two-element lists: the first to "turn on" some feature of the
markup language, and the second to turn it off.  For example, if your
target language is LaTeX2e and you want journal names emphasized, you
would supply a list reference C<['\emph{','}']> for the C<journal_mkup>
option.  If you were instead generating HTML, you might supply
C<['E<lt>emphE<gt>','E<lt>/emphE<gt>']>.  To keep the structure module
general with respect to markup languages, all markup options are empty
by default.  (Or, rather, they are all references to lists consisting of
two empty strings.)

=over 4

=item C<name_mkup>

Markup to add around the list of author names.

=item C<atitle_mkup>

Markup to add around non-book (article) titles.

=item C<btitle_mkup>

Markup to add around book titles.

=item C<journal_mkup>

Markup to add around journal names.

=back

=cut

my %default_options =
   (sortby      => 'name',              # or 'year', 'none'
    namestyle   => 'full',              # or 'abbrev', 'nopunct', 'nospace'
    nameorder   => 'first',             # or 'last'
    atitle_lower=> 1,                   # mangle case of non-book titles?
    labels      => 'numeric',           # or 'alpha' (not yet supported!)
    name_mkup   => ['', ''],
    atitle_mkup => ['', ''],
    btitle_mkup => ['', ''],
    journal_mkup=> ['', ''],
   );


=head2 Option methods

As required by the C<Text::BibTeX::Structure> module,
C<Text::BibTeX::Bib> provides two methods for handling options:
C<known_option> and C<default_option>.  (The other two option methods,
C<set_options> and C<get_options>, are just inherited from
C<Text::BibTeX::Structure>.)

=over 4

=item known_option (OPTION)

Returns true if OPTION is one of the options on the above list.

=item default_option (OPTION)

Returns the default value of OPTION, or C<croak>s if OPTION is not a
supported option.

=back

=cut

sub known_option 
{
   my ($self, $option) = @_;
   return exists $default_options{$option};
}


sub default_option
{
   my ($self, $option) = @_;
   return exists $default_options{$option}
      ? $default_options{$option}
      : $self->SUPER::default_option ($option);
}


# The field lists in the following documentation are automatically
# generated by my `doc_structure' program -- I run it and read the
# output into this file.  Wouldn't it be cool if the module could just
# document itself?  Ah well, dreaming again...

=head1 DATABASE STRUCTURE

The other purpose of a structure class is to provide a method,
C<describe_entry>, that lists the allowed entry types and the known
fields for the structure.  Programmers wishing to write their own
database structure module should consult L<Text::BibTeX::Structure> for
the conventions and requirements of this method; the purpose of the
present document is to describe the C<Bib> database structure.

The allowed entry types, and the fields recognized for them, are:

=over 4

=item C<article>

Required fields: C<author>, C<title>, C<journal>, C<year>.
Optional fields: C<volume>, C<number>, C<pages>, C<month>, C<note>.

=item C<book>

Required fields: C<title>, C<publisher>, C<year>.
Optional fields: C<series>, C<address>, C<edition>, C<month>, C<note>.
Constrained fields: exactly one of C<author>, C<editor>; at most one of C<volume>, C<number>.

=item C<booklet>

Required fields: C<title>.
Optional fields: C<author>, C<howpublished>, C<address>, C<month>, C<year>, C<note>.

=item C<inbook>

Required fields: C<publisher>, C<year>.
Optional fields: C<series>, C<type>, C<address>, C<edition>, C<month>, C<note>.
Constrained fields: exactly one of C<author>, C<editor>; at least one of C<chapter>, C<pages>; at most one of C<volume>, C<number>.

=item C<incollection>

Required fields: C<author>, C<title>, C<booktitle>, C<publisher>, C<year>.
Optional fields: C<editor>, C<series>, C<type>, C<chapter>, C<pages>, C<address>, C<edition>, C<month>, C<note>.
Constrained fields: at most one of C<volume>, C<number>.

=item C<inproceedings>

=item C<conference>

Required fields: C<author>, C<title>, C<booktitle>, C<year>.
Optional fields: C<editor>, C<series>, C<pages>, C<address>, C<month>, C<organization>, C<publisher>, C<note>.
Constrained fields: at most one of C<volume>, C<number>.

=item C<manual>

Required fields: C<title>.
Optional fields: C<author>, C<organization>, C<address>, C<edition>, C<month>, C<year>, C<note>.

=item C<mastersthesis>

Required fields: C<author>, C<title>, C<school>, C<year>.
Optional fields: C<type>, C<address>, C<month>, C<note>.

=item C<misc>

Required fields: none.
Optional fields: C<author>, C<title>, C<howpublished>, C<month>, C<year>, C<note>.

=item C<phdthesis>

Required fields: C<author>, C<title>, C<school>, C<year>.
Optional fields: C<type>, C<address>, C<month>, C<note>.

=item C<proceedings>

Required fields: C<title>, C<year>.
Optional fields: C<editor>, C<series>, C<address>, C<month>, C<organization>, C<publisher>, C<note>.
Constrained fields: at most one of C<volume>, C<number>.

=item C<techreport>

Required fields: C<author>, C<title>, C<institution>, C<year>.
Optional fields: C<type>, C<number>, C<address>, C<month>, C<note>.

=item C<unpublished>

Required fields: C<author>, C<title>, C<note>.
Optional fields: C<month>, C<year>.

=back 

=cut

sub describe_entry
{
   my $self = shift;

   # Advantages of the current scheme (set all fields for a particular
   # entry type together):
   #   - groups fields more naturally (by entry type)
   #   - might lend itself to structuring things by 'type' in the object
   #     as well, making it easier to determine if a type is valid
   #   - prevents accidentally giving a type optional fields but no
   #     required fields -- currently this mistake would make the type
   #     'unknown'
   # 
   # Requirement of any scheme:
   #   - must be easy for derived classes to override/augment the field
   #     lists defined here! (ie. they should be able just to inherit 
   #     describe_entry; or explicitly call SUPER::describe_entry and then
   #     undo/change some of its definitions

   # Things that I don't think are handled by this scheme, but that
   # bibtex does look out for:
   #  * warns if month but no year
   #  * crossref stuff:
   #    - article can xref article; xref'd entry must have key or journal
   #    - book or inboox can xref book; xref'd entry must have editor,
   #      key, or series
   #    - incollection can xref a book and inproceedings can xref a 
   #      proceedings; xref'd entry must have editor, key, or booktitle

   $self->set_fields ('article',
                      [qw(author title journal year)],
                      [qw(volume number pages month note)]);
   $self->set_fields ('book',
                      [qw(title publisher year)],  
                      [qw(series address edition month note)],
                      [1, 1, [qw(author editor)]],
                      [0, 1, [qw(volume number)]]);
   $self->set_fields ('booklet',
                      [qw(title)],
                      [qw(author howpublished address month year note)]);
   $self->set_fields ('inbook',
                      [qw(publisher year)],
                      [qw(series type address edition month note)],
                      [1, 1, [qw(author editor)]],
                      [1, 2, [qw(chapter pages)]],
                      [0, 1, [qw(volume number)]]);
   $self->set_fields ('incollection',
                      [qw(author title booktitle publisher year)],
                      [qw(editor series type chapter pages address 
                          edition month note)],
                      [0, 1, [qw(volume number)]]);
   $self->set_fields ('inproceedings',
                      [qw(author title booktitle year)],
                      [qw(editor series pages address month 
                          organization publisher note)],
                      [0, 1, [qw(volume number)]]);
   $self->set_fields ('conference',
                      [qw(author title booktitle year)],
                      [qw(editor series pages address month 
                          organization publisher note)],
                      [0, 1, [qw(volume number)]]);
   $self->set_fields ('manual',
                      [qw(title)],
                      [qw(author organization address edition 
                          month year note)]);
   $self->set_fields ('mastersthesis',
                     [qw(author title school year)],
                     [qw(type address month note)]);
   $self->set_fields ('misc',
                      [],
                      [qw(author title howpublished month year note)]);
   $self->set_fields ('phdthesis',
                      [qw(author title school year)],
                      [qw(type address month note)]);
   $self->set_fields ('proceedings',
                      [qw(title year)],
                      [qw(editor series address month 
                          organization publisher note)],
                      [0, 1, [qw(volume number)]]);
   $self->set_fields ('techreport',
                      [qw(author title institution year)],
                      [qw(type number address month note)]);
   $self->set_fields ('unpublished',
                      [qw(author title note)],
                      [qw(month year)]);

}  # describe_entry


=head1 STRUCTURED ENTRY CLASS

The second class provided by the C<Text::BibTeX::Bib> module is
C<BibEntry> (again, a leading C<Text::BibTeX> is implied).  This being a
structured entry class, it derives from C<StructuredEntry>.  The
conventions and requirements for such a class are documented in
L<Text::BibTeX::Structure> for the benefit of programmers implementing
their own structure modules.

If you wish to write utilities making use of the C<Bib> database
structure, then you should call one of the "officially supported"
methods provided by the C<BibEntry> class.  Currently, there are only
two of these: C<sort_key> and C<format>.  These are actually implemented
in the C<BibSort> and C<BibFormat> classes, respectively, which are base
classes of C<BibEntry>.  Thus, see L<Text::BibTeX::BibSort> and
L<Text::BibTeX::BibFormat> for details on these two methods.

=cut

package Text::BibTeX::BibEntry;
use strict;
use vars qw(@ISA);

use Text::BibTeX::BibSort;
use Text::BibTeX::BibFormat;

@ISA = qw(Text::BibTeX::BibSort Text::BibTeX::BibFormat);

# Pre-define the "month name" macros for compatibility with BibTeX.
# This ignores all sorts of issues, like internationalization and
# abbreviation.  
my %month_names = 
   ('jan' => 'January',
    'feb' => 'February',
    'mar' => 'March',
    'apr' => 'April',
    'may' => 'May',
    'jun' => 'June',
    'jul' => 'July',
    'aug' => 'August',
    'sep' => 'September',
    'oct' => 'October',
    'nov' => 'November',
    'dec' => 'December');

my ($macro, $expansion);
Text::BibTeX::add_macro_text ($macro, $expansion)
   while (($macro, $expansion) = each %month_names);    

1;

=head1 SEE ALSO

L<Text::BibTeX::Structure>, L<Text::BibTeX::BibSort>, 
L<Text::BibTeX::BibFormat>.

=head1 AUTHOR

Greg Ward <gward@python.net>

=head1 COPYRIGHT

Copyright (c) 1997-2000 by Gregory P. Ward.  All rights reserved.  This file
is part of the Text::BibTeX library.  This library is free software; you
may redistribute it and/or modify it under the same terms as Perl itself.
